pkg/kubelet/kubelet.go

```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
 	kl.syncLoop(updates, kl)
 }
 
+func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
+	node := &api.Node{
+		ObjectMeta: api.ObjectMeta{Name: kl.hostname},
+		Status: api.NodeStatus{
+			Capacity: api.ResourceList{
+				api.ResourceCPU:    *resource.NewMilliQuantity(1000, resource.DecimalSI),
+				api.ResourceMemory: resource.MustParse("3Gi"),
+			},
+		},
+	}
+	if kl.cloud != nil {
+		instances, ok := kl.cloud.Instances()
+		if !ok {
+			return nil, fmt.Errorf("failed to get instances from cloud provider")
+		}
+		// TODO(roberthbailey): Inspect the local machine rather than calling out
+		// to the cloud provider.
+		resources, err := instances.GetNodeResources(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get node resources from cloud provider: %v", err)
+		}
+		if resources != nil {
+			node.Status.Capacity = resources.Capacity
+		}
+		instanceID, err := instances.ExternalID(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
+		}
+		node.Spec.ExternalID = instanceID
+		nodeAddresses, err := instances.NodeAddresses(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get node address from cloud provider: %v", err)
+		}
+		node.Status.Addresses = nodeAddresses
+	}
+	return node, nil
+}
+
+// registerNode registers the node with the cluster master.
+func (kl *Kubelet) registerNode() error {
+	step := 100 * time.Millisecond
+	for start := time.Now(); time.Since(start) < 30*time.Minute; time.Sleep(step) {
+		step = step * 2
+		if step >= 7*time.Second {
+			step = 7 * time.Second
+		}
+
+		node, err := kl.initialNodeStatus()
+		if err != nil {
+			glog.Errorf("Unable to construct api.Node object for kubelet: %v", err)
+			continue
+		}
+		glog.V(2).Infof("Attempting to register node %s", node.Name)
+		if _, err := kl.kubeClient.Nodes().Create(node); err != nil {
+			glog.V(2).Infof("Unable to register %s with the cluster master: %v", node.Name, err)
+			continue
+		}
+		glog.Infof("Successfully registered node %s", node.Name)
+		return nil
+	}
+	return errors.New("unable to register node with the master after 30 minutes")
+}
+
 // syncNodeStatus periodically synchronizes node status to master.
 func (kl *Kubelet) syncNodeStatus() {
 	if kl.kubeClient == nil {
 		return
 	}
+	if err := kl.registerNode(); err != nil {
+		glog.Errorf("Failed to register node: %s. Giving up.", err)
```

Comment:
I didn't realize that existed -- cool. I've passed it a new error so that we keep the custom text on top of the error text returned from registerNode.